### PR TITLE
EL-4232 - Option to prevent menu from closing on item click/enter

### PR DESCRIPTION
**Pull request by cmallon367**
#### Checklist

* [x] The contents of this PR are mine to submit. No third party code or other material is being committed.
* [x] New third party dependencies (if any) are linked via `package.json`. Third party dependencies are licensed under one of the following: Apache, MIT, BSD, Mozilla Public License, Eclipse Public License, or Oracle Binary Code License.
* [x] The code in this PR does not contain export-restricted encryption algorithms.

#### Ticket / Issue

https://portal.digitalsafe.net/browse/EL-4232

#### Description of Proposed Changes

Added new input 'closeOnSelect' and updated the new input on the documentation site
Added karma tests for the new input
#### Documentation CI URL

https://pages.github.houston.softwaregrp.net/sepg-docs-qa/UXAspects_CI_UXAspects_EL-4232-prevent-menu-close-on-item-click

#### Downstream Build(s)

* [caf/ux-aspects-micro-focus](https://jenkins.swinfra.net/job/SEPG/job/caf/job/caf~ux-aspects-micro-focus~EL-4232-prevent-menu-close-on-item-click~CI/)

### CODE REVIEW

File: src/components/menu/menu-item/menu-item.component.ts [hunk 1]
```diff
@@ -127,7 +138,7 @@ export class MenuItemComponent implements OnInit, OnDestroy, FocusableOption {
     @HostListener('click', ['$event'])
     @HostListener('keydown.enter', ['$event'])
     _onClick(event: MouseEvent | KeyboardEvent): void {
-        if (!this.disabled) {
+        if (!this.disabled && this.closeOnSelect) {
             this.onClick$.next(isKeyboardTrigger(event) ? 'keyboard' : 'mouse');
             this.activate.emit(event);
```
Comment (JonathanToddMF) [inline src/components/menu/menu-item/menu-item.component.ts:143]: The above could be the cause of this, we are stopping the activate from being emitted here. Maybe we could have another if statement just for `closeOnSelect` which stops the `this.onClick$.next` but keeps the `this.closeOnSelect`?

**Review by JonathanToddMF — CHANGES_REQUESTED** [review]
Looks good and has solved the issue although there appears to be a regression

I have added the new input to the `Mark As` menu item and its submenus `Pending`, `Escalated `and `Closed` in the following plunker - https://plnkr.co/edit/ZnI7oNcL285BmI5C

You can see that the submenu does not close when the options are selected.

However, when using `closeOnSelect` with the `activate` output it no longer seems to emit when it is clicked on.  If you check the plunker I have provided, removing the `closeOnSelect` causes the activate to work as intended.

**Review by JonathanToddMF — DISMISSED** [review]
Looks good Christina, the `activate` output is now working with `closeOnSelect`

Fix detailed here - https://plnkr.co/edit/ZnI7oNcL285BmI5C

You can see the activate output working in the console. Note Mark As > Closed does not have the `closeOnSelect` applied to test this still works without it.

**Review by donna-marijne — CHANGES_REQUESTED** [review]
Can you add some tests?

File: docs/app/pages/components/components-sections/buttons/dropdowns/dropdowns.component.html [hunk 2]
```diff
@@ -210,6 +210,9 @@ <h4>Flat</h4>
     <tr uxd-api-property name="role" type="'menuitem' | 'menuitemradio' | 'menuitemcheckbox'" defaultValue="menuitem">
         Define the role of the menu item.
     </tr>
+    <tr uxd-api-property name="closeOnSelect" type="boolean" defaultValue="true">
+        Determine if the menu should close on click of the submenu.
```
Comment (donna-marijne) [inline docs/app/pages/components/components-sections/buttons/dropdowns/dropdowns.component.html:214]:
```suggestion
        Determine if the menu should close on click of the menu item.
```

**Review by donna-marijne — CHANGES_REQUESTED** [review]
Looks good, I just noticed one error in the documentation.

**Review by donna-marijne — APPROVED** [review]
Looks good, thank you.

**Review by donna-marijne — DISMISSED** [review]
Looks good, thank you.

**Review by JonathanToddMF — DISMISSED** [review]
Tests look good, QA to follow

**Comment by JonathanToddMF** [comment]
QA: One minor issue found

Following plunker - https://plnkr.co/edit/pCgTHupy4op8zOky - shows the new `closeOnSelect` input on the `Pending` and `Escalated` sub menu items inside `Mark as`. Clicking on these keeps the menu open and does not close it as required. `Closed` inside `Mark as` does not have this new input and closes correctly.

Defects:
When navigating the menu with the keyboard controls pressing enter on one of the items that has `closeOnSelect` set as true causes the `activate` output to trigger twice

File: src/components/menu/menu-item/menu-item.component.ts [hunk 3]
```diff
@@ -127,8 +138,15 @@ export class MenuItemComponent implements OnInit, OnDestroy, FocusableOption {
     @HostListener('click', ['$event'])
     @HostListener('keydown.enter', ['$event'])
     _onClick(event: MouseEvent | KeyboardEvent): void {
+        if (event instanceof MouseEvent) {
+            if (event.screenX === 0 && event.screenY === 0) {
+                return;
+            }
+        }
```
Comment (JonathanToddMF) [inline src/components/menu/menu-item/menu-item.component.ts:145]: We should probably combine these if statements

`if (event instanceof MouseEvent && event.screenX === 0 && event.screenY === 0) {
    return;
}`

**Review by JonathanToddMF — CHANGES_REQUESTED** [review]
Hi Christina, all looks good with this although I would combine the if statement you've added.

**Review by JonathanToddMF — APPROVED** [review]
preventDefault is definitely a lot clearer. Approved thanks!

**Comment by JonathanToddMF** [comment]
Qa complete and working

**Comment by JonathanToddMF** [comment]
QA complete and tested in the following browsers - Chrome, Firefox, Edge and IE11

Following plunker - https://plnkr.co/edit/pCgTHupy4op8zOky - shows the new closeOnSelect input on the Pending and Escalated sub menu items inside Mark as. Clicking on these keeps the menu open and does not close it as required. Closed inside Mark as does not have this new input and closes correctly.

Using keyboard controls no longer emits the activate output twice.

Awaiting acceptance @Roland-Penny

**Review by Roland-Penny — APPROVED** [review]
Looks good.

**Pull request merged**